custom_components/sonnenbatterie/sensor.py

```diff
@@ -61,517 +58,68 @@ def _internal_setup(_username, _password, _ipaddress):
     )
 
+    async_add_entities(
+        SonnenbatterieSensor(coordinator=coordinator, entity_description=description)
+        for description in generate_powermeter_sensors(_coordinator=coordinator)
+    )
 
-class SonnenBatterieSensor(CoordinatorEntity, SensorEntity):
-    def __init__(self, entity_id, device_info, coordinator, name=None):
-        self._attributes = {}
-        self._state = "0"
-        self._device_info = device_info
-        self.coordinator = coordinator
-        self.entity_id = entity_id
-        if name is None:
-            name = entity_id
-        self._name = name
-        super().__init__(coordinator)
-        LOGGER.info("Create Sensor {0}".format(entity_id))
-
-    @property
-    def device_info(self) -> DeviceInfo:
-        """Return the device info."""
-        return self._device_info
-
-    def set_state(self, state):
-        """Set the state."""
-        if self._state == state:
-            return
-        self._state = state
-        if self.hass is None:
-            # LOGGER.warning("hass not set, sensor: {} ".format(self.name))
-            return
-        self.schedule_update_ha_state()
-        # try:
-        # self.schedule_update_ha_state()
-        # except:
-        #    LOGGER.error("Failing sensor: {} ".format(self.name))
-
-    def set_attributes(self, attributes):
-        """Set the state attributes."""
-        self._attributes = attributes
-
-    @property
-    def unique_id(self) -> str:
-        """Return the unique ID for this sensor."""
-        return self.entity_id
-
-    @property
-    def should_poll(self):
-        """Only poll to update phonebook, if defined."""
-        return False
-
-    @property
-    def device_state_attributes(self):
-        """Return the state attributes."""
-        return self._attributes
-
-    @property
-    def state(self):
-        """Return the state of the device."""
-        return self._state
-
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return self._name
-
-    def update(self):
-        LOGGER.info("update " + self.entity_id)
-
-    @property
-    def unit_of_measurement(self):
-        """Return the unit of measurement."""
-        return self._attributes.get("unit_of_measurement", None)
-
-    @property
-    def device_class(self):
-        """Return the device_class."""
-        return self._attributes.get("device_class", None)
+    LOGGER.info("Init done")
+    return True
 
-    @property
-    def state_class(self):
-        """Return the unit of measurement."""
-        return self._attributes.get("state_class", None)
 
+class SonnenbatterieSensor(CoordinatorEntity[SonnenBatterieCoordinator], SensorEntity):
+    """Represent an SonnenBatterie sensor."""
 
-class SonnenBatterieCoordinator(DataUpdateCoordinator):
-    """My custom coordinator."""
+    entity_description: SonnenbatterieSensorEntityDescription
+    _attr_should_poll = False
+    _attr_has_entity_name = True
```

Comment:
Wenn du das `_attr_has_entity_name = True` weglässt hast du vermutlich weniger Probleme mit den Entity-Namen. Zumindest hat das Setzen dieses Attribute bei meinen Tests bewirkt, dass HA sich immer irgendwelche komischen Sachen für den Namen ausgewürfelt hat.

Reply:
Also ich hatte jetzt nicht das Gefühl, dass die Namen "komisch" sind. Die Entities heißen doch genau so wie in der translation angegeben + noch den Gerätename davor. Das der Gerätename davor steht und dann in der Geräteübersicht automatisch ausgeblendet wird ist aber auch in den offiziellen Integrationen so, ist also wohl so gewollt. Ich persönlich finde das jetzt auch nicht schlecht.